doc/dev/design/NAT-address-discovery.rst

```diff
@@ -0,0 +1,143 @@
+as an extension to the discover/bootstrapping service.
+
+Proposal
+========
+[A precise statement of the proposed change.]
+
+Proposed change: Extend the border router (BR) to detect NATed addresses/ports and report them back to a client endhost.
+
+Ideally the solution would listen on the same port that would also be used for normal traffic forwarding.
+Alternatively, it may be feasible (depending on how sensitive the NAT is) to use a different (fixed) port to
+accept request for reporting addresses.
+
+The implementation on the protocol level could be done in several ways:
+
+1.  STUN: The BR needs to detect if an incoming packet is a STUN packet, and if it is, treat it is such and respond
+    to the STUN request.
+
+    -  Advantages: STUN is a well known and mature protocol. There are STUN libraries available in many programming languages.
+       It should be easy to implement this.
+
+2.  Extend SCMP with a new INFO message
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: More standardization effort? How do we solve authentication?
+
+3.  Extend SCMP with a new ERROR message: "invalid source address for first hop pkt", similar to error 33.
+    The router can verify that for first hop packets, the IP src address/port matches the SCION src address/port.
+    If not, it returns an error. The actual source address would need to be attached somewhere, unless we decide
+    to change the payload so it contains the IP header of the offending packet (and the IP header should contain the
+    NATed IP/port).
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: Conceptually a bit of a hack. More standardization effort? How do we solve authentication?
+
+
+Rationale
+=========
+[A discussion of alternate approaches and the trade-offs, advantages, and disadvantages of the specified approach.]
+
+The main reasons for integrating the functionality with the BR are:
+
+-  Reliability: The border router is almost guaranteed to see the correct IP/port on the NAT, especially when using the
+   same port for STUN traffic and routing traffic. All other approaches rely on the leniency of the NAT to use the same
+   port even if the STUN server and border router have different ports or even IPs.
+-  Time to rollout: changing the border routers should be much easier and faster than getting router vendors or NAT
+   vendors to implement SCION compatibility or to get rid of NATs completely in home networks.
+   A short time until rollout seems important because people are already running into this problem.
+
+Alternatives:
+
+-  The SRC address/port is updated by the border router to reflect what the border router sees as source address.
+   Problem:
+   -  Complicates cryptographically protecting the header if the header must be modifiable by border routers
```

Comment:
Refer to SPAO [https://docs.scion.org/en/latest/protocols/authenticator-option.html]